csrc/ir_base_nodes.cpp

```diff
@@ -112,6 +112,33 @@ const std::vector<Expr*>& Val::uses() const {
   return uses_;
 }
 
+bool Val::addUse(Expr* expr, bool defer_inserting_tv_uses) {
```

Comment:
I'm feeling this is becoming more complex than necessary. Why not just make `addUse` add a use expr, and whether invalidating or not should be decided by `Fusion`?

I think all we would need to do is to change `registerExpr` to call `addUse` only when the `Val` is not tensor. I added a suggestion there.